examples/02_model_collaborative_filtering/cornac_bpr_deep_dive.ipynb

```diff
@@ -88,16 +88,38 @@
+    "\n",
+    "<p align=\"center\">\n",
+    "<img src=\"https://recodatasets.z20.web.core.windows.net/images/bpr1.png\" width=\"400px\">\n",
+    "</p>\n",
+    "\n",
+    "In the implicit feedback scenario the observed data is positive feedback. As it is shown in the figure above, the observed data (image to the left) is known, but the unobserved data is unknown. The negative data is usually generated by filling the matrix with 0 (image to the right). \n",
+    "\n",
+    "Using the pointwise loss, the model is trained to predict as 1 the elements in the observed dataset and 0 for the rest. The problem with this approach is that all elements the model should rank in the future are presented as negative feedback during training. Therefore, an overfitted model, a model that learns the training data exactly, will predict all new elements as 0s. The only reason these models can predict rankings is by using techniques to prevent overfitting, like regularization.\n",
+    "\n",
+    "BPR uses a different approach by using item pairs $(i, j)$ and optimizing for the correct ranking given preference of user $u$, thus, there are notions of *positive* and *negative* items. The training data $D_S : U \\times I \\times I$ is defined as:\n",
     "\n",
     "$$D_S = \\{(u, i, j) \\mid i \\in I^{+}_{u} \\wedge j \\in I \\setminus I^{+}_{u}\\}$$\n",
     "\n",
     "where user $u$ is assumed to prefer $i$ over $j$ (i.e. $i$ is a *positive item* and $j$ is a *negative item*).\n",
     "\n",
+    "This approach uses the **pairwise loss**. The intuition behind it is that observed entries should be ranked higher than the unobserved ones. Then instead of minimizing the loss between $\\hat{x}_{u,i}$ and ${x}_{u,i}$, pairwise learning maximizes the margin between the observed entry $\\hat{x}_{u,i}$ and the unobserved entry $\\hat{x}_{u,j}$.\n",
+    "\n",
+    "<p align=\"center\">\n",
+    "<img src=\"https://recodatasets.z20.web.core.windows.net/images/bpr2.png\" width=\"400px\">\n",
+    "</p>\n",
+    "\n",
+    "In the figure above, pairwise learning is explained. For each user, a pairwise matrix of items is generated. The plus indicates that the user prefers item *i* over item *j*, and the minus is that the user prefers item *j* over *i*. The interrogation indicates that we don't know the user's preference.\n",
+    "\n",
+    "The pairwise setup has several advantages:\n",
+    "* The training set is defined as positive, negative and unknown preferences. \n",
+    "* The missing values between two non-observed items are exactly the item pairs that have to be ranked by the model.\n",
+    "* The training data is created for the actual objective of ranking.\n",
     "\n",
```

Comment:
Btw the pairwise loss of BPR is not specific to implicit feedback. It could easily be applied when the data contains ratings, since they can be ordered ("user u prefers movie a more than movie b").

Reply:
yeah, that's a good point. Maybe we can add another point in the line of:

* Pairwise ranking is not restricted to implicit feedback and can be used in explicit feedback by setting user preferences as pairs (i.e. user u prefers item a more than item b).

What do you think?

Reply:
Yes, maybe also link to Cornac or some paper that does this.

Reply:
changed